internal/heartbeat/heartbeat.go

```diff
@@ -219,19 +230,20 @@ func (s *Heartbeat) Init(config models.DeviceConfigurationMessage) error {
 }
 
 func (s *Heartbeat) Update(config models.DeviceConfigurationMessage) error {
-	s.tickerLock.Lock()
+	s.tickerLock.RLock()
 	periodSeconds := s.getInterval(*config.Configuration)
 	if s.previousPeriodSeconds <= 0 || s.previousPeriodSeconds != periodSeconds {
 		s.log.Debugf("Heartbeat configuration update: periodSeconds changed from %d to %d; Device ID: %s", s.previousPeriodSeconds, periodSeconds, s.data.workloadManager.GetDeviceID())
-		s.tickerLock.Unlock()
+		// Had to unlock here to avoid deadlock when call s.initTicker
+		s.tickerLock.RUnlock()
 		s.log.Infof("reconfiguring ticker with interval: %v. DeviceID: %s", periodSeconds, s.data.workloadManager.GetDeviceID())
 		if s.ticker != nil {
 			s.ticker.Stop()
 		}
 		s.initTicker(periodSeconds)
```

Comment:
maybe here, the lock should happens inside the initticker and have a function called stopTicker that does lock/unlock for simplicity

Reply:
I used atomic to avoid having to `RLock` then manually (meaning no `defer`) ` when checking the value of `previousPeriodSeconds`  then RUnlock`, `Lock` and manually `Unlock` when setting it (I have to be inside the condition) 
Using the atomic allows me avoid all this

Reply:
One lock for entire update was not enough?